tf2_ros/include/tf2_ros/message_filter.h

```diff
@@ -95,14 +99,19 @@ static std::string get_filter_failure_reason_string(
   filter_failure_reasons::FilterFailureReason reason)
 {
   switch (reason) {
-    case filter_failure_reasons::Unknown:
-      return "Unknown";
     case filter_failure_reasons::OutTheBack:
-      return "OutTheBack";
+      return
+        "the timestamp on the message is more than the cache length earlier than the newest data"
+        "in the transform cache";
```

Comment:
I think this should read:
```suggestion
        "the timestamp on the message is more than the cache length and earlier than the newest data"
        "in the transform cache";
```
?

Also, it seems like this case is never happening... :thinking:

Reply:
yes, I can delete that case if desired, it isn't being used anywhere (except [this](https://github.com/ros2/rviz/blob/9696867179460b68ee90292f4ad89207280a4798/rviz_common/src/rviz_common/frame_manager.cpp#L308) commented out code).

> I think this should read:

I'm not sure. I copied that from the error docblock and IMO it means:

"the timestamp on the message is earlier than all the data in the transform cache"

but I really cannot check what was the meaning because there's nothing triggering that error condition :confused:.

Reply:
I found the case in ROS 1's `tf2::BufferCore`: https://github.com/ros/geometry2/blob/c73b5939723db078c9bbe18523230ad54f859682/tf2/src/buffer_core.cpp#L1286-L1297

Yeah, you are correct. It happens when the timestamp is too old (out of the cache).

In ROS 1, this case was handled in the message filter: https://github.com/ros/geometry2/blob/c73b5939723db078c9bbe18523230ad54f859682/tf2_ros/include/tf2_ros/message_filter.h#L331-L336

But in ROS 2, the only place I can find handling this case is in in `tf2_ros::Buffer`:

https://github.com/ros2/geometry2/blob/f4ea23490cbbdde8c18725a458153a5fe5d456e7/tf2_ros/src/buffer.cpp#L256-L262

Reply:
Thanks for looking into the details.
So that "OutTheBack" error should be set now [here](https://github.com/ros2/geometry2/blob/f4ea23490cbbdde8c18725a458153a5fe5d456e7/tf2_ros/include/tf2_ros/message_filter.h#L518), but I'm still not sure what to do with [this](https://github.com/ros2/geometry2/blob/f4ea23490cbbdde8c18725a458153a5fe5d456e7/tf2_ros/include/tf2_ros/message_filter.h#L529) and [this](https://github.com/ros2/geometry2/blob/f4ea23490cbbdde8c18725a458153a5fe5d456e7/tf2_ros/include/tf2_ros/message_filter.h#L538).

I think we can keep the unused "OutTheBack" error for the moment, and try to use it in a follow-up improvement.

Reply:
Ok, I think this is what the ROS 1 code was doing https://github.com/ros2/geometry2/pull/364/commits/9462264ac3e8d675d31cbd29c3ddb3b21d94ab72.

Reply:
SGTM